### DISCUSSION TITLE
Range Rant

### DISCUSSION THREAD

**Discussion by I-Omnibus** [dotnet/csharplang#4933] [Language Ideas]
```csharp
/// <summary>
/// <see cref="System.Diagnostics.CodeAnalysis.Range"/>
/// 
/// The compiler Range syntax ('start..end') is too good to be restricted to just 'int' types.
/// 
/// <code>public readonly struct Range<T> where T : struct</code>
/// 
/// Perhaps the compiler could recognise this pattern for generic types, making it more generally available,  
/// to simplify and make code more readable?
/// 
/// The Start and End properties would have to be Nullable<T> to overcome an issue where '..^0' cannot be
///  distinguished from '0..^0' and likewise, '0..' is indistinguishable from '0..^0' which makes drawing an
///  inference from an absent Start or End impossible.
/// 
/// </summary>

using System;
using Oversimplification;

/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
/*                                                           */
/*                   var range = 'a'..'z';                   */
/*                                                           */
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
var range = new DotDotRangeSyntax<char>(start: 'a', end: 'z');
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

namespace Oversimplification {
	interface IRange<T> where T : struct {
		IIndex<T> Start { get; init; }
		IIndex<T> End { get; init; }
		string ToString();
	}
	interface IIndex<T> where T : struct {
		Nullable<T> Value { get; init; }
		bool FromEnd { get; init; }
		string ToString();
	}

	readonly struct DotDotRangeSyntax<T> : IRange<T> where T : struct {
		public DotDotRangeSyntax(DotDotIndex<T> start, DotDotIndex<T> end)
			=> (Start, End) = (start, end);
		public IIndex<T> Start { get; init; }
		public IIndex<T> End { get; init; }
		public override string ToString() => $"{Start}..{End}";
	}

	readonly struct DotDotIndex<T> : IIndex<T> where T : struct {
		public DotDotIndex(Nullable<T> value, bool fromEnd = false)
			=> (Value, FromEnd) = (value, fromEnd);
		public Nullable<T> Value { get; init; }
		public bool FromEnd { get; init; }
		public static implicit operator DotDotIndex<T>(Nullable<T> value)
			=> new(value);
		public override string ToString() => $"{Value}";
	}
}
```

**Comment by qrli** [comment]
My understanding is that: the range feature is really "Index Range" but not other ranges.
Other ranges have very different expectation and handling, which may be too specific to bake into the language. E.g. 
* A float point range's Start and End are typically inclusive, because their Start & End are infinite small point thus `range.Length == range.End - range.Start`. While index range typically have End exclusive, to satisfy `range.Length == range.End - range.Start`. There also range with Infinite and NaN values, where different libraries do not agree upon their behaviors
* For ranges used in graphics, there is no default start or end typically, unlike indexes where there is an implicit Length/Count. You cannot do `1..`, nor `1..^3`. However, graphics range typically supports reversed ranges, e.g. `90..10`.
* Depends on context, some range types may also have operations, like Include, Exclude, Shrink, Expand, etc., which may not apply to other ranges, or have different expectations on behavior.

**Reply by jnm2** [reply]
> Depends on context, some range types may also have operations, like Include, Exclude, Shrink, Expand, etc., which may not apply to other ranges, or have different expectations on behavior.

Seems like an ideal fit for extension methods.

**Comment by ArcanoxDragon** [comment]
I just stumbled across this discussion after (disappointedly) learning that the `x..y` operator can't be overloaded. It seems like this concept was implemented in a very restrictive manner compared to other language features.

Suppose there's a `CellAddress` type representing the absolute address of one cell on an Excel-styled spreadsheet. Suppose there's also a `CellRange` type representing the rectangular region between two `CellAddress`es. The `CellRange` type could implement `IEnumerable<CellAddress>` to enumerate through all individual cells inside the range.

What's stopping the language from supporting something like this?

```csharp
public static CellRange operator ..(CellAddress left, CellAddress right)
```

This would lend itself to something like this:

```csharp
foreach (CellAddress address in topLeft .. bottomRight) { ... }
```
